### PR TITLE
Require attributes to derive from System.Attribute

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
This fixes https://github.com/Microsoft/visualfsharp/issues/4460

image

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
buildfromsource.cmd does not see the new resource string:
```
D:\j\w\release_ci_pa---c6929ad7\src\fsharp\TypeChecker.fs(10658,130): error FS0039: The field, constructor or member 'tcTypeDoesNotInheritAttribute' is not defined. 
```
How to fix this?

**Comment by smoothdeveloper** [comment]
@vasily-kirichenko you need to copy fscomp* files under src/buildfromsource/FSharp.Compiler.Private

**Comment by smoothdeveloper** [comment]
see https://github.com/Microsoft/visualfsharp/blob/master/DEVGUIDE.md#updating-fscompfs

**Comment by vasily-kirichenko** [comment]
@smoothdeveloper thanks you! done.

**Comment by vasily-kirichenko** [comment]
It's ready.

**Comment by vasily-kirichenko** [comment]
I've merged master into it. It's still ready :)

File: src/buildfromsource/FSharp.Compiler.Private/FSComp.resx [hunk 1]
```diff
@@ -4354,4 +4354,7 @@
   <data name="chkNoSpanLikeValueFromExpression" xml:space="preserve">
     <value>A Span or IsByRefLike value returned from the expression cannot be used at ths point. This is to ensure the address of the local value does not escape its scope.</value>
   </data>
+  <data name="tcTypeDoesNotInheritAttribute" xml:space="preserve">
+    <value>This type is not an attribute</value>
```
Comment (cartermp) [inline src/buildfromsource/FSharp.Compiler.Private/FSComp.resx]: I'd suggest changing it to:

"This type is not an attribute. To make it an attribute, you must derive from System.Attribute"

Comment (vasily-kirichenko) [reply src/buildfromsource/FSharp.Compiler.Private/FSComp.resx]: I copied it from the C# error message. I don't think longer message is better.

**Comment by vasily-kirichenko** [comment]
> Since this is a breaking change, we'd likely not take this until after 4.5 is released, and only in F# 5.0.

It's not a breaking change since using non attribute type as attribute results with runtime exceptions today. It's a (rather important) bug fix.

File: src/buildfromsource/FSharp.Compiler.Private/FSComp.resx (hunk 1 above)
Comment (cartermp) [reply src/buildfromsource/FSharp.Compiler.Private/FSComp.resx]: It was done correctly from the beginning in C# by having this restriction. We're now introducing a restriction that previously wasn't there, which I think justifies a suggestion in the error message.

**Comment by cartermp** [comment]
This is by definition a source-breaking change, regardless of if the current behavior can result in an issue at runtime. I'm 100% 👍  with doing this, but only for F# 5.0 (which should correspond with VS 2019 RTW). We (Microsoft) will also be responsible for documenting this difference in behavior between 4.5 and 5.0 and appropriately messaging this.

**Comment by abelbraaksma** [comment]
I'd personally have this in rather sooner than later, but perhaps meanwhile we can have a highest level warning instead of an error, plus message of deprecation? Than at least people can cherry pick this feature with warnings as errors.

**Comment by vasily-kirichenko** [comment]
I still don't see how it's a breaking change. If a user has ever used a non-attribute type as attribute, he saw runtime errors and has already fixed such mistakes. So, there are zero (used by anybody) codebases which has such mistakes, and this is not a breaking change.

**Comment by matthid** [comment]
Question is at what time exactly it is failing at runtime? when loading the assembly? when retrieving attributes via reflection? when the type is loaded?

**Comment by vasily-kirichenko** [comment]
@matthid actually it does not throw at all

image

**Comment by vasily-kirichenko** [comment]
And you can even use structs :)

image

**Comment by matthid** [comment]
Then it is a breaking change ;)

**Comment by matthid** [comment]
But what we should do now is insert a warning. can we do that @cartermp?

**Comment by cartermp** [comment]
I don't think it is likely that this change will be able to make it into F# 4.5 or a point release afterwards, especially given:

- We're locked down on F# 4.5, with the _current_ policy being only critical bug fixes and big regressions being allowed beyond the 15.8 release of VS
- Our recent history of breaking changes in a point release (tuple aliasing, inheriting from FSharpFunc, SRTP changes)

I wouldn't even be comfortable with a warning until 5.0, since it's likely that a lot of systems in the world are compiling with warnings as errors.

**Comment by forki** [comment]
@cartermp we should not start to see "warnings as errors" as something that is related to breaking/non breaking decisions. That will make bad precedent.

**Comment by cartermp** [comment]
Perhaps we should re-frame the discussion on this PR.

Aside from consistency with C#, what exactly does this fix? And is that worth taking a breaking change for?

**Comment by matthid** [comment]
> Aside from consistency with C#, what exactly does this fix? And is that worth taking a breaking change for?

I'd say it breaks a general assumption of developers, which can lead to subtle errors. Also there are APIs which return an array of System.Attribute, correct? what will happen with those?

**Comment by matthid** [comment]
Regular APIs returning System.Attribute like https://msdn.microsoft.com/en-us/library/w11a4f55(v=vs.110).aspx will fail, correct?

**Comment by cartermp** [comment]
`GetCustomAttribute(s)` APIs don't fail for this:

```fsharp
open System

[<AttributeUsage(AttributeTargets.All, Inherited=false, AllowMultiple=true)>]
type Author(name: string) =
    member val Version = 1.0 with get, set
    member __.GetName() = name

[<Author("P. Ackerman")>]
type FirstClass() = class end

type SecondClass() = class end

[<Author("P. Ackerman"); Author("R. Koch", Version=2.0)>]
type ThirdClass = class end

let print(t: Type) =
    let attrs = Attribute.GetCustomAttributes(t) 
    for attr in attrs do
        printfn "%A" attr

[<EntryPoint>]
let main argv =
    print typeof<FirstClass>
    print typeof<SecondClass>
    print typeof<ThirdClass>
    0 // return an integer exit code
```

However, I don't think you can actually get the underlying information from the `Author` class unless you derive from `Attribute`. Casting is a compile error already.

Which is why I'm interested in what the runtiume _failure_ mentioned by @eiriktsarpalis in the original isue was. Because I don't think the .NET runtime actually requires something to inherit from `Attribute` for it to be treated as an attribute.

**Comment by KevinRansom** [comment]
@cartermp 

This is how you get the author data.  Use GetCustomAttributes() from the type object rather than the Attribute type.

````
open System

[<AttributeUsage(AttributeTargets.All, Inherited=false, AllowMultiple=true)>]
type Author(name: string) =
    member val Version = 1.0 with get, set
    member __.GetName() = name

[<Author("P. Ackerman")>]
type FirstClass() = class end

type SecondClass() = class end

[<Author("P. Ackerman"); Author("R. Koch", Version=2.0)>]
type ThirdClass = class end

let print(t: Type) =
    let attrs = t.GetCustomAttributes(false) 
    for attr in attrs do
            if attr.GetType().IsAssignableFrom(typeof<Author>) then printfn "%s" ((attr :?> Author).GetName())

[<EntryPoint>]
let main argv =
    print typeof<FirstClass>
    print typeof<SecondClass>
    print typeof<ThirdClass>
    0 // return an integer exit code
````


Produces:
````
P. Ackerman
P. Ackerman
R. Koch


````

**Comment by KevinRansom** [comment]
@vasily-kirichenko , @cartermp 

Clearly this would be a breaking change, we could argue among ourselves about how common not using an Attribute base class is, I would say that most developers who also use C# would automatically add the Attribute base class.  Exclusive F# developers may be less inclined to, however the Attribute APIs would likely also push them towards using the Attribute base class.

Because this is a breaking change we certainly want existing projects to continue to compile.  However, the interop with C#/VB and some of the .NET Attribute APIs, we need to do something to notify F# developers of the issue.

I would suggest that we make this a warning with error text along the lines of:
__"This type does not inherit Attribute, it will not work correctly with other .NET languages."__.

And for new projects we make this a compile error.

In existing projects, developers will see the warning, and can decide whether to address the error or disable the warning.

In new projects with the dotnet cli, and legacy .fsproj files they will see it as an error, and so the old code will hopefully bitrot away.

As for the when:  the F# 4.5 ship has sailed, but I think this seems a safe enough change for Dev 16.0

**Comment by KevinRansom** [comment]
@vasily-kirichenko, @cartermp, @dsyme 

@vasily it looks like you have done the hard work, here should I take over and finish it off, or do you want to finish the work to productise this?

Kevin

**Comment by vasily-kirichenko** [comment]
@KevinRansom Please, finish it.

**Comment by KevinRansom** [comment]
@vasily-kirichenko thanks mate, will do.  And thanks for driving this so far, from inside the bubble here it's hard to find these kind of fit and finish issues.

**Comment by abelbraaksma** [comment]
> Because I don't think the .NET runtime actually requires something to inherit from Attribute for it to be treated as an attribute.

@cartermp, to be blunt (CLI Annotated Standard also says, just before that (Partition IIA, section 20)): _"CLS compliance requires that attributes will be instances of types whose parent is `System.Attribute`"_

But, just before that: _"... any user-defined type can be used as an attribute"_

Furthermore CLS Rule 41 says: _"Attributes shall be of type `System.Attribute` or inherit from it."_, with the following notes:

CLS (consumer): need not support attributes that are not of the specified type.
CLS (extender): must support the authoring of custom attributes.
CLS (framework): shall not publicly expose attributes that are not type `System.Attribute` or a type inheriting from it.

The way I understand this is: code that consumes attributes that do not inherit from `System.Attribute` is not CLS compliant and any other consumer, need not support it. The framework _does not_ expose those.

I don't know how strict these rules are, but I think a (strong) warning that attributes cannot be consumed unless they inherit from `System.Attribute` seems to me the least we should do.

As an aside: ending the name of an attribute with `Attribute` suffix is a guideline, but not a requirement. There are also some rules on disambiguating fringe cases where the short name is used, but a class exists with/without the `Attribute` suffix, but that's a bit outside of the scope of this issue (and I did not test how F# handles these fringe cases).

**Comment by KevinRansom** [comment]
this will be handled by this PR: https://github.com/Microsoft/visualfsharp/pull/5610

**Pull request closed without merging**